### DISCUSSION TITLE
@urql/Svelte query needs to return promise

### DISCUSSION THREAD

**Discussion by dhinesh03** [urql-graphql/urql#2373] [Q&A] [answered]
I need to get data from the graphql server on my component's onMount lifecycle function, after getting the response I have to set it on my 3rd party plugin. I couldn't find any solution on @urql/Svelte docs. So I wrote a small wrapper for that. It's working properly.  Please suggest to me is there any other better way of handling this.

```
<script> 
import { onMount, tick } from "svelte";
import { query, operationStore } from "@urql/svelte";
import Choices from 'choices.js';
let tagsElement;

const tags = operationStore(`query { tags { id name} } }`, undefined, { pause: true });
query(tags)

async function executeQuery(operationStore) {
      operationStore.reexecute({ requestPolicy: "network-only" });
      let unsub;
      return new Promise((resolve, reject) => {
          unsub = operationStore.subscribe(() => {
              if (operationStore.data) {
                  resolve(operationStore.data);
              } else if (operationStore.error) {
                  reject(operationStore.error);
              }
          });
      }).finally(() => {
          unsub();
          console.log("Experiment completed");
      });
}

onMount(()=>{
    const tagsElementWithChoices = new Choices(tagsElement);
    const response = await executeQuery(tags);
    tagsElementWithChoices.setChoices(response.tags, 'id', 'name');
})
</script>
<input class="tags" type="text" bind:this={tagsElement} on:change={onTagsChange} />
```
@frederikhors

**Comment by freb** [comment, marked as answer]
I haven't tested this much, but I think I finally figured out how to create a promise that resolves to a store with the v2 Svelte bindings.

If you're using SvelteKit, it is nice to use promises in the `load` function to guarantee that data is available when the page gets rendered. One way to do this is with the urql client directly: `client.query().toPromise()`. This returns a "one-time" result, and not a store, so it is not reactive. You can use `queryStore` to get a store, but then you cannot wait for the request to complete in the load function.

I'm using the following to return a promise that resolves to a store, ensuring that the request has completed before it resolves. There may be some edge cases that I have not run into yet that were not an issue when using `toPromise()` directly, but it seems to be working.

I'm using graphql-codegen to generate my types, so you can see my helpers for my `standards` node type.

~~~ts
// Create a store for a query of standards.
export const standardsStore = (vars: StandardsQueryVariables = {}, client: Client = get(clientStore), requestPolicy: RequestPolicy | undefined = undefined) => (
	queryStore<StandardsQuery, StandardsQueryVariables>({
		client: client,
		query: standardsQuery,
		variables: vars,
		requestPolicy: requestPolicy,
	})
)

// Create a promise that resolves to a store for a query of standards.
export const standards = (vars: StandardsQueryVariables = {}, client: Client = get(clientStore), requestPolicy: RequestPolicy | undefined = undefined) => {
	return new Promise((resolve, reject) => {
		const s = standardsStore(vars, client, requestPolicy)
		s.subscribe(value => {
			if (!value.fetching || value.error) {
				resolve(s)
			}
		})
	})
}
~~~
